cpp/src/arrow/util/basic_decimal.cc

```diff
@@ -248,40 +247,50 @@ BasicDecimal128& BasicDecimal128::operator>>=(uint32_t bits) {
   return *this;
 }
 
-BasicDecimal128& BasicDecimal128::operator*=(const BasicDecimal128& right) {
-  // Break the left and right numbers into 32 bit chunks
-  // so that we can multiply them without overflow.
-  const uint64_t L0 = static_cast<uint64_t>(high_bits_) >> 32;
-  const uint64_t L1 = static_cast<uint64_t>(high_bits_) & kIntMask;
-  const uint64_t L2 = low_bits_ >> 32;
-  const uint64_t L3 = low_bits_ & kIntMask;
+namespace {
+
+void extendAndMultiplyUint64(uint64_t x, uint64_t y, uint64_t* hi, uint64_t* lo) {
+  const uint64_t x_lo = x & kIntMask;
```

Comment:
I think we can add an optimization:

```
#ifdef __SIZEOF_INT128__
  // use __uint128_t
#else
  // fallback to current logic
#endif
```
Please also run the micro benchmark a few times (the results of the first few runs after compilation are usually noisy).

Reply:
Done.

Previous results:
BinaryMathOp                          137 ns          137 ns 
BinaryMathOpAggregate          143 ns          143 ns

New results:
BinaryMathOp                          135 ns          135 ns
BinaryMathOpAggregate          143 ns          143 ns

I also performed some micro benchmarks on my own which show the current multiplication method takes ~3.2 ns, __int128 multiplication takes ~0.7 ns and the new method can take between ~2.3 and ~3.2 ns depending on how many negations must occur.